### ISSUE TITLE
modifying the `rankingScoreThreshold` changes the results' ranking 

### ISSUE THREAD

**Issue by CaroFG** [meilisearch/meilisearch#4988]
**Describe the bug**

Given the same query, changing the `rankingScoreThreshold` modifies the search result list.
Also  [reported](https://discord.com/channels/1006923006964154428/1292804022948073522/1292804022948073522) by a Discord member and on GH: #4989 

**To Reproduce**
I use the movies.json dataset 
1. search for "spring" with a 0.96 rankingScoreThreshold:
```
{
    "q": "spring",
    "showRankingScore": true,
    "rankingScoreThreshold": 0.96
}
```
Returns the following hits:

```
 {
            "id": 241855,
            "title": "Spring",
            "overview": "A young man in a personal tailspin flees the US to Italy, where he sparks up a romance with a woman harboring a dark, primordial secret.",
            "genres": [
                "Horror",
                "Romance",
                "Science Fiction"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/qPI94JmMga88RB7vuQA83McgFaz.jpg",
            "release_date": 1412726400,
            "_rankingScore": 1.0
        },
        {
            "id": 113,
            "title": "Spring, Summer, Fall, Winter... and Spring",
            "overview": "An isolated lake, where an old monk lives in a small floating temple. The monk has a young boy living with him, learning to become a monk. We watch as seasons and years pass by.",
            "genres": [
                "Drama"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/6SQQ5REuAz7k0FMQ9mSCT40T2LN.jpg",
            "release_date": 1063929600,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 16394,
            "title": "Winnie the Pooh: Springtime with Roo",
            "overview": "Spring has sprung, and baby Roo is excited to get out and explore and make new friends. But Rabbit seems preoccupied with spring cleaning, instead of embracing his usual role of playing Easter Bunny. Leave it to Roo to show Rabbit -- through love -- that it's more important who you love and not who's in charge.",
            "genres": [
                "Animation",
                "Family"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/yAgVcU1BdPqvnyYnOkb1QtVNd5V.jpg",
            "release_date": 1078790400,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 18826,
            "title": "Spring Breakdown",
            "overview": "Three women in their thirties head to a popular Spring Break destination and try to relive the Spring Break they never had.",
            "genres": [
                "Comedy"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/g9wU95xHvsWl7ZRMhUeXBOkpExY.jpg",
            "release_date": 1242950400,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 21148,
            "title": "Spring Break",
            "overview": "Two sets of two college guys spend a spring break together in Fort Lauderdale, Florida. There they have lots of fun in and out of the sun.",
            "genres": [
                "Comedy",
                "Romance"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/dShsZsu3YOccxaORvm1pXsxCaPZ.jpg",
            "release_date": 417398400,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 27647,
            "title": "Spring Break Massacre",
            "overview": "When six sorority sex kittens hook up for a night of carefree indulgence during spring break, they have no idea that maniacal serial killer Stanley Peterson (Curtis Taylor) has escaped from a nearby prison and is now on the prowl for fresh blood. Can they stop the madman before he notches up another half-dozen deadly conquests?",
            "genres": [
                "Horror"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/2cyaCjxqO5FVhDvhaVzXT6qjguC.jpg",
            "release_date": 1207008000,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 44732,
            "title": "Spring Break Shark Attack",
            "overview": "A young teenager travels to Florida, unknowing that a group of dangerous tiger sharks are ravaging the beach.",
            "genres": [
                "Action",
                "Drama",
                "Horror",
                "Thriller"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/jqIIwaYI4to46Cc29as0BCECqNO.jpg",
            "release_date": 1111276800,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 51145,
            "title": "Spring Forward",
            "overview": "Trying to put himself back together after serving prison time for robbery, Paul takes a job as a parks employee in a small Connecticut town. His co-worker, Murph, a generation older than Paul, is dealing with his own issues, chiefly the the impending death of his very ill son. After the men have a run-in with local snob Frederickson, Murph's age and experience starts to temper Paul's fiery nature, as the younger man begins to embrace his new life.",
            "genres": [
                "Drama"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/id2pAgtG5dNS7lguTpXbhCq9l3g.jpg",
            "release_date": 976406400,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 122081,
            "title": "Spring Breakers",
            "overview": "After four college girls rob a restaurant to fund their spring break in Florida, they get entangled with a weird dude with his own criminal agenda.",
            "genres": [
                "Drama",
                "Crime"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/9tyPnyEkL44qbAliM9jMRWc6bjg.jpg",
            "release_date": 1346803200,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 199905,
            "title": "Spring Thaw",
            "overview": "Katelyn 'Kate' Weston, a country girl who became a lawyer and joined an L.A. law firm years ago, specializes in defending minors having committed violent crimes who nobody else takes on. She gradually got estranged from her family, but then her older brother Ryan asks her to return to the farm he runs with their dad Thane who has become impossibly absentminded. By the time he is formally diagnosed with serious Alzheimer, she realizes how badly she really knew her family, having ran from her past after the trauma of witnessing her mother's suicide at age seven, and tries to talk it trough with Ryan and dad while he still has lucid moments.",
            "genres": [
                "Drama",
                "TV Movie"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/qBaICuRkAj051uu3lDvMhu5VdwP.jpg",
            "release_date": 1167609600,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 323465,
            "title": "After Dark",
            "overview": "Spring break for seven college students is derailed with one wrong turn. A flat tire on a deserted road leaves them stranded after dark in the company of a menacing drifter who may or may not be trying to help.",
            "genres": [
                "Horror"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/h3pggCoU8NUuz3cmhbKCLkMatcd.jpg",
            "release_date": 1422921600,
            "_rankingScore": 0.9848484848484849
        },
        {
            "id": 199782,
            "title": "Springsteen & I",
            "overview": "For 40 years Bruce Springsteen has influenced fans from all over. His songs defined more than a generation. This film gives the fans just as much time as The Boss himself, with never shown footage and live performances from his last tour.",
            "genres": [
                "Documentary",
                "Music"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/4EfNnkusEyw5fQYvHEGKMnrf7m1.jpg",
            "release_date": 1374451200,
            "_rankingScore": 0.9621212121212122
        }
```

2. Raise the threshold to 0.99:
```
{
    "q": "spring",
    "showRankingScore": true,
    "rankingScoreThreshold": 0.99
}
```
Returns the following results:

```
        {
            "id": 241855,
            "title": "Spring",
            "overview": "A young man in a personal tailspin flees the US to Italy, where he sparks up a romance with a woman harboring a dark, primordial secret.",
            "genres": [
                "Horror",
                "Romance",
                "Science Fiction"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/qPI94JmMga88RB7vuQA83McgFaz.jpg",
            "release_date": 1412726400,
            "_rankingScore": 1.0
        },
        {
            "id": 9041,
            "title": "One Deadly Summer",
            "overview": "In spring 1976, a 19-year-old beauty, her German-born mother, and her crippled father move to the town of a firefighter nicknamed Pin-Pon. Everyone notices the provocative Eliane. She singles out Pin-Pon and soon is crying on his shoulder (she's myopic and hates her reputation as a dunce and as easy); she moves in with him, knits baby clothes, and plans their wedding. Is this love or some kind of plot? She asks Pin-Pon's mother and aunt about the piano in the barn: who delivered it on a November night in 1955? Why does she want to know, and what does it have to do with her mother's sorrows, her father's injury, this quick marriage, and the last name on her birth certificate?",
            "genres": [
                "Mystery",
                "Drama"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/cVShoNWphKc5txc3sfl6Uf8OU92.jpg",
            "release_date": 421459200,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 20530,
            "title": "Late Spring",
            "overview": "Noriko is perfectly happy living at home with her widowed father, Shukichi, and has no plans to marry -- that is, until her aunt Masa convinces Shukichi that unless he marries off his 27-year-old daughter soon, she will likely remain alone for the rest of her life. When Noriko resists Masa's matchmaking, Shukichi is forced to deceive his daughter and sacrifice his own happiness to do what he believes is right.",
            "genres": [
                "Drama"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/iNtRSY2AGjW1VDXDR79bKsNUdus.jpg",
            "release_date": -640656000,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 43674,
            "title": "The Spring",
            "overview": "A man and his son stumble onto a modern-day fountain of youth while camping in the woods.",
            "genres": [
                "Science Fiction"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/pZc7Daa0et3YBzs3p3g7ZvHLjbl.jpg",
            "release_date": 947980800,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 92662,
            "title": "The Hollywood Complex",
            "overview": "Every spring, Hollywood hosts a very species-specific migration: kids. Thousands of aspiring child actors flock to Tinseltown for pilot season, the traditional casting period for new network and cable television shows. But unlike adult actors who pound the same star-lined pavement, kids come with their families. Many set up camp at the Oakwood, a temporary housing complex that caters to the showbiz flock. While their little Angelinas and mini Brads audition for armies of ruthless agents and TV executives, the parents must manage sky-high expectations and thin skins, including their own",
            "genres": [
                "Documentary",
                "Comedy",
                "Family"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/4ygFG7NjrnmVZLEt82qr7noCkHg.jpg",
            "release_date": 1304208000,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 114574,
            "title": "Bikini Spring Break",
            "overview": "A group of coeds from a small conservative college break out of their shells when their marching band bus breaks down in Fort Lauderdale during Spring Break.",
            "genres": [
                "Comedy"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/uYpQnHfFqPRxOdEFExs9xHPNrXu.jpg",
            "release_date": 1340668800,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 207345,
            "title": "1313: Haunted Frat",
            "overview": "During Spring Break, a fraternity built atop the burned ruins of an insane asylum is haunted by a sinister visitor...",
            "genres": [
                "Thriller",
                "Horror"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/eJerQjijN7YHAGpxjxubAfJnSDw.jpg",
            "release_date": 1312243200,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 208512,
            "title": "Blast Vegas",
            "overview": "Thieving Spring Breakers steal an ancient relic that unleashes a disastrous curse upon Las Vegas.  Stars: Frankie Muniz, Maggie Castle, and Barry Bostwick",
            "genres": [
                "Thriller",
                "TV Movie",
                "Science Fiction",
                "Action",
                "Adventure"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/AoK8wcJmCKy0l2pWBPzssqve3bi.jpg",
            "release_date": 1374019200,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 272078,
            "title": "Coldplay: Ghost Stories - Live 2014",
            "overview": "In Spring 2014, Coldplay returned with their sixth studio album 'Ghost Stories'. The band unveiled the album to a small handful of fans via an intimate, world-premiere performance at Sony Studios, L.A. This ground-breaking production was performed in the round with jaw-dropping 360-degree and overhead imagery. Director Paul Dugdale explored worlds-within-worlds, blending live performance with cinematic film.",
            "genres": [
                "Music"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/qvmBzdLwfAYstwbp1EV9WaOTgjx.jpg",
            "release_date": 1400457600,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 283727,
            "title": "Blue Spring Ride",
            "overview": "When she was a middle school student, Futaba Yoshioka liked Kou Tanaka who was unlike the other boys and quite gentle. Kou Tanaka then moved away and they lost contact.  At the end of Futaba Yoshioka's first year as a high school student she meets Kou Tanaka again. Kou Tanaka’s name has changed to Kou Mabuchi and his personality seems different from his middle school days.",
            "genres": [
                "Romance",
                "Drama"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/8ic3JNoQVPTGtOwgUeTWu09mmNl.jpg",
            "release_date": 1418428800,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 292532,
            "title": "Wicked Spring",
            "overview": "During the 1864 battle of the Wilderness, three Union soldiers and three Confederate Soldiers get seperated from their units as twilight engulfs the ravaged battlefield. The men wander alone through the dangerous woods, separate of each other, until they meet by chance on the banks of a quiet creek. The men meet and spend the night around a campfire, not realizing they are enemies until the next morning when the sun rises and a new day of battle begins.",
            "genres": [
                "Western",
                "Drama",
                "War"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/s9qtTzz7zcV6A9eMAz8Ml74TRON.jpg",
            "release_date": 1015027200,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 400408,
            "title": "Highway to Havasu",
            "overview": "During spring break, two buddies kidnap their broken-hearted friend and embark on a road trip to Lake Havasu.",
            "genres": [
                "Comedy"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/zzjx1PCmSWUnHsjPJptJj6T3Bhs.jpg",
            "release_date": 1488844800,
            "_rankingScore": 0.9931129476584022
        },
        {
            "id": 21544,
            "title": "Hope Springs",
            "overview": "British artist Colin Ware discovers that his fiancée, Vera, is going to marry another man. Distraught and despondent, he gets on a plane for America and ends up in the tiny town of Hope in New England. At first, Colin is depressed, but he soon finds more than a shoulder to cry on when his innkeepers introduce him to Mandy, a beautiful nurse. All's going well and Colin has almost forgotten his old flame until Vera shows up with a surprise of her own.",
            "genres": [
                "Comedy",
                "Romance"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/oqYtfJYApyVHzFwS5YCPpRBTZs4.jpg",
            "release_date": 1052438400,
            "_rankingScore": 0.9924242424242424
        },
        {
            "id": 37536,
            "title": "Ringmaster",
            "overview": "Jerry Springer stars as more or less himself, the host of a raunchy, controversial and popular Los Angeles talk show which features everyday people with problems and who frequently vent their problems on the air.",
            "genres": [
                "Comedy"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/9QNqeKB4XodrvEiYewOTG9WEka5.jpg",
            "release_date": 911952000,
            "_rankingScore": 0.9924242424242424
        },
        {
            "id": 37826,
            "title": "Bruce Springsteen and the E Street Band : Live in New York City",
            "overview": "A two disc amalgam of the final performances of 2001's Madison Square Gardens performances by one of the greatest bands in the world of some of the greatest music in the world. The atmosphere positively floods out of the screen to envelop you and the hairs on your neck will be standing on end before the first note has been struck. After watching this you'll believe that The Boss is incapable of putting a foot wrong. By the end, he's only just short of defying gravity.",
            "genres": [
                "Music"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/4dusouw2kGsvfvjMhFMTUVXBt5a.jpg",
            "release_date": 962236800,
            "_rankingScore": 0.9924242424242424
        },
        {
            "id": 49417,
            "title": "Bruce Springsteen & the E Street Band – London Calling Live in Hyde Park",
            "overview": "Bruce Springsteen &amp; The E Street Band's London Calling: Live In Hyde Park concert film is of the Boss and his band's performance at London's Hard Rock Calling Festival on June 28, 2009 in HD. The film documents 27 tracks of live Springsteen that begin in daylight and progress through a gorgeous sunset into night. The set list spans from the Born To Run era to Working On a Dream and includes rare covers such as The Clash's \"London Calling\", Jimmy Cliff's \"Trapped\", The Young Rascals' \"Good Lovin'\", and Eddie Floyd's \"Raise Your Hand\". Springsteen also performs fan favorite \"Hard Times (Come Again No More)\". Brian Fallon from The Gaslight Anthem joins the band as a guest vocalist on Springsteen's own \"No Surrender\".",
            "genres": [
                "Music"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/cUQ6h9IcXpniPFsrmgHE0KPZyUV.jpg",
            "release_date": 1246147200,
            "_rankingScore": 0.9924242424242424
        },
        {
            "id": 64533,
            "title": "Bruce Springsteen & the E Street Band Houston '78 Bootleg: House Cut",
            "overview": "Previously unreleased complete 3-hour concert performance from Bruce Springsteen's historic Darkness on the Edge of Town Tour filmed in Houston in 1978. This release is titled \"Bootleg: House Cut\" as it is the actual footage that appeared on-screen at the concert.",
            "genres": [
                "Music"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/r7ZMkyy6NJ216DRjAQVzkXEjTHS.jpg",
            "release_date": 1289865600,
            "_rankingScore": 0.9924242424242424
        },
        {
            "id": 82696,
            "title": "Hope Springs",
            "overview": "After thirty years of marriage, a middle-aged couple attends an intense, week-long counseling session to work on their relationship.",
            "genres": [
                "Drama",
                "Comedy",
                "Romance"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/mHEca5NntVHtUPfBiQ3ZVFBRixe.jpg",
            "release_date": 1344297600,
            "_rankingScore": 0.9924242424242424
        },
        {
            "id": 83186,
            "title": "Wrong",
            "overview": "Dolph Springer wakes up one morning to realize he has lost the love of his life, his dog, Paul. During his quest to get Paul (and his life) back, Dolph radically changes the lives of others -- risking his sanity all the while.",
            "genres": [
                "Comedy",
                "Mystery"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/y2leAYKv7fMs3ZjngZY6HJjZTkE.jpg",
            "release_date": 1346803200,
            "_rankingScore": 0.9924242424242424
        },
        {
            "id": 229702,
            "title": "Zephyr Springs",
            "overview": "When Dawn, an overworked single mother, agrees to take her teenage daughter, Kayla, to a luxurious but isolated spa retreat, the women believe they have escaped to a paradise of spa treatments, yoga, and nature hikes until a series of disturbing incidents has Kayla questioning the perfection around her.",
            "genres": [
                "Thriller"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/6I4h9A9wKscN86LKmSyZ8iil4HE.jpg",
            "release_date": 1370044800,
            "_rankingScore": 0.9924242424242424
        }
```

**Expected behavior**
I would expect the second query with the 0.99 threshold to return the first result of the first request. I don't expect to have different search results with different ranking scorers for the same query.

**Screenshots**

**Meilisearch version:**
1.10.2

**Additional context**
Additional information that may be relevant to the issue.
[e.g. architecture, device, OS, browser]

**Comment by dureuill** [comment]
Thanks for reporting, both you and @zperzendetta 

I reproduced the bug with `movies.json` in the default settings and the two queries:
```jsonc
{
    "q": "spring",
    "showRankingScore": true,
    "rankingScoreThreshold": 0.96
}
```

and 
```jsonc
{
    "q": "spring",
    "showRankingScore": true,
    "rankingScoreThreshold": 0.99 // raised to 0.99
}
```

Enabling the `rankingScoreDetails` with `showRankingScoreDetails: true` indicates that the `rankingScore` of e.g. `"One Deadly Summer"` does not agree with its `rankingScoreDetails`. Trying with a lower rankingScoreThreshold (0.93) and a high limit, we can find `'One Deadly Summer"` later in the list of results, with a more expected `rankingScore` of `0.9242424242424242`.

There is definitely a bug here, and it seems related to the computation of the `rankingScore` from the `rankingScoreDetails` when the `rankingScoreThreshold` is enabled. I'll discuss the priority of the issue with the team.

**Comment by teodorstupnicki** [comment]
Hi @curquiza, I'm already proficient in Rust but I'm new to the codebase. Any tips on how to start?

**Comment by dureuill** [comment]
hey @teodorstupnicki , the logic to compute the ranking score and decide whether to keep the document is in https://github.com/meilisearch/meilisearch/blob/main/milli/src/search/new/bucket_sort.rs#L240.

**Comment by teodorstupnicki** [comment]
@dureuill Thank you! Can you assign this to me? I'll give it a shot

**Comment by curquiza** [comment]
Thanks @teodorstupnicki for your interest in this project!

For your information, we prefer not assigning people to our issues because sometimes people ask to be assigned and never come back, which discourages the volunteer contributors from opening a PR to fix this issue.
We will accept and merge the first PR that fixes correctly and well implements the issue following our [contributing guidelines](../blob/main/CONTRIBUTING.md).   

We are looking forward to reviewing your PR 😊

**Comment by teodorstupnicki** [comment]
@curquiza Sure, I understand. Sounds like a reasonable policy 😊

**Comment by Gnosnay** [comment]
looks interesting. i also wanna have one try on this fix. ;)

**Comment by curquiza** [comment]
Hello @Gnosnay 

Thanks for your interest in this project 🔥 You are definitely more than welcome to open a PR for this!

For your information, we prefer not assigning people to our issues because sometimes people ask to be assigned and never come back, which discourages the volunteer contributors from opening a PR to fix this issue.
We will accept and merge the first PR that fixes correctly and well implements the issue following our [contributing guidelines](../blob/main/CONTRIBUTING.md).   

We are looking forward to reviewing your PR 😊

**Comment by dureuill** [comment]
Hello, our [BENCHMARKS.md](https://github.com/meilisearch/meilisearch/blob/main/BENCHMARKS.md) file might help with that

**Comment by Gnosnay** [comment]
> Hello, our [BENCHMARKS.md](https://github.com/meilisearch/meilisearch/blob/main/BENCHMARKS.md) file might help with that, but to troubleshoot this issue you mostly need the movies.json _dataset_ rather than _workload_
> 
> (dataset is the data, here the movie descriptions and titles, workloads are sequences of commands that are sent to a Meilisearch instance to execute automatically and reproducibly)

thx! @dureuill i also found the doc from https://www.meilisearch.com/docs/learn/self_hosted/getting_started_with_self_hosted_meilisearch

BENCHMARKS.md also helped a lot. thank you for your sharing!

**Comment by Gnosnay** [comment]
hi there, i dont think this is one good first issue.. i require more help to fix this.

i tried to use step debugger to find out the reason why we got different scores, i make one break point above `maybe_add_to_results`

and i found that
- when the `rankingScoreThreshold` is 0.92, then `ranking_rule_scores` will have 7 elements
- when the `rankingScoreThreshold` is 0.99, then `ranking_rule_scores` will have 8 elements
here are the difference:

0.92:
<img width="686" alt="image" src="https://github.com/user-attachments/assets/f863d597-8e6e-4817-b723-ef78df14f924" />

0.99:
<img width="706" alt="image" src="https://github.com/user-attachments/assets/eb11d1b2-ceac-466c-a369-1f2ba8eb9603" />

for 0.99 one, therer is one more `ScoreDetail::Position` with `rank = 11` and `max_rank = 11`.

I believe it is related to how we generate details in `bucket`. I am not really familiar with this part. i will very appreciate it if anyone can give some clues. thx!

**Comment by Gnosnay** [comment]
hey guys, please kindly let me know if anyone can give some brief about this part.

Keywords are okay to me, i can do some invetstigation about this

**Comment by irevoire** [comment]
Hey @Gnosnay, I quickly looked into it but don't really know the answer, sorry.
@dureuill will be more knowledgeable than me on this part, but he's on holiday for the week. I'll ping him so he sees this message once he comes back. Thanks for your interest in this issue :pray:

**Comment by Gnosnay** [comment]
> Hey @Gnosnay, I quickly looked into it but don't really know the answer, sorry.
> 
> @dureuill will be more knowledgeable than me on this part, but he's on holiday for the week. I'll ping him so he sees this message once he comes back. Thanks for your interest in this issue :pray: 

Hi @irevoire oh it is okay! No need to feel sorry! Thank you so much for your reply.

Let us wait for a bit and happy holidays!

**Comment by dureuill** [comment]
hello, 

indeed that one is a bit on the harder side of "good first issue" 😳

as you might have found, the logic happens in `crates/milli/src/search/new/bucket_sort.rs`, and precisely in the `bucket_sort` function.

the `bucket_sort` function applies the ranking rules one by one and [gets a bucket of results](https://github.com/meilisearch/meilisearch/blob/ba11121cfc822438659ccb4120327c0c211a2796/crates/milli/src/search/new/bucket_sort.rs#L216). That bucket also contains the score details of the applied ranking rule for the documents it contains. 
The score details are pushed to `ranking_rule_scores` which contains all the score details for the successive ranking rules that were applied to return the bucket. Then, when there's a ranking score threshold, the [`_rankingScore` so far is computed](https://github.com/meilisearch/meilisearch/blob/ba11121cfc822438659ccb4120327c0c211a2796/crates/milli/src/search/new/bucket_sort.rs#L241) and the documents are not added to results and removed from candidates if their current score is under the threshold.
 
As a note, the logic operates under the assumption that the global score is decreasing when adding a score details to a prefix of score details, that is, given the score details A, B and C,

```
global_score([A, B]) <= global_score([A, B, C])
```

---


From your investigation, it looks like the issue is that the proximity ranking rule is somehow applied twice to the document when the 0.99 score is returned. This is faulty, and the next step would be to understand why `position` is pushed twice in the logic of the `bucket_sort` function. This can be done by monitoring pushes to the `ranking_rule_scores` variable.


---

Lastly, a detail: your debugger screenshots are beautiful; which IDE are you using?

**Comment by Gnosnay** [comment]
i think i got the reason but i don't understand it really clearly. Here is my conclusion:

1.  When score is 0.99, for the first iteraion of `while valid_docids.len() < length {` loop, it will get the same `ranking_rule_scores` as 0.96
<img width="831" alt="image" src="https://github.com/user-attachments/assets/0b4f5c61-3579-40c1-bcfd-c4f656ed086c" />

2. but after that, it hits the condition https://github.com/meilisearch/meilisearch/blob/9c1e54a2c86b1d39db975fc90bace2affaa5fa8a/crates/milli/src/search/new/bucket_sort.rs#L218
3. then https://github.com/meilisearch/meilisearch/blob/9c1e54a2c86b1d39db975fc90bace2affaa5fa8a/crates/milli/src/search/new/bucket_sort.rs#L223 will remove the last score
4. then the loop will hit https://github.com/meilisearch/meilisearch/blob/9c1e54a2c86b1d39db975fc90bace2affaa5fa8a/crates/milli/src/search/new/bucket_sort.rs#L172 and it will trigger `back!()` which will reduce `cur_ranking_rule_index` by 1 and pop the last element of `ranking_rule_scores` again.
5. so before the iteration begins really, the last element will be the `Position`, and the `cur_ranking_rule_index` is 5.
6. then iteration begins, the `ExactAttribute` is added back again. this time it hits the https://github.com/meilisearch/meilisearch/blob/9c1e54a2c86b1d39db975fc90bace2affaa5fa8a/crates/milli/src/search/new/bucket_sort.rs#L208 which will call `back!()` again. so `cur_ranking_rule_index` will be deducted by 1 again. then it will be `4` which will add one more `Position` in next iteration.

**Comment by Gnosnay** [comment]
TBH i don't really understand the logic about this whole part. so i can not give accurate dignostics. if you could reply, i would really appreciate it.

## what things are sorted in bucket sort?

As far as i know, it seems that we will nomalize the documents in some ways by using `RoaringBitmap`, which actually is `universe: &RoaringBitmap` in `bucket_sort`.

And it is generated from `[crates/milli/src/search/new/mod.rs:567](https://github.com/meilisearch/meilisearch/blob/525e67ba93c64f8d8d9856573d0d3bc23a2a2b80/crates/milli/src/search/new/mod.rs#L567)`. But i know few about this.

So for this part my questions are:
- What is `universe`?
- How is the basic logic about generation of `universe`?
- How is the document mapped to bit in `universe`?
- What is the relation between `docid` in `universe` and id listed in response of doc?

## what is main idea of `bucket_sort`? and what is the bucket?

It seems that the target of `bucket_sort` is getting some valid docids in the end. but there are some details here i could not understand.

it is true that we use `ScoreDetails`, which is `ranking_rule_scores` in `bucket_sort`,  to calculate the `current_score` by some algorithm in `ScoreDetails::global_score`.

But what i could not understand are:
- what is the bucket? coz `ranking_rule_scores` is populated by `next_bucket.score`, and the valid doc ids are calculated by deducting the `next_bucket.candidates`. But i know few about this core structure.
- from the code https://github.com/meilisearch/meilisearch/blob/9c1e54a2c86b1d39db975fc90bace2affaa5fa8a/crates/milli/src/search/new/bucket_sort.rs#L169 we can only end the loop when
    - get enough valid docids
    - time is out
   but why? i don't really understand the reason here, because it seems that the bug appears because we can not stop the loop and somehow the loop will deduct the scores array which will add same `Position` back.
- why did we design one `back!()` function to remove the scores and re-add them again to get more documents?

**Comment by Gnosnay** [comment]
Lastly, a detail: your debugger screenshots are beautiful; which IDE are you using?
---

thank you for your curiority, my IDE is from JetBrain which is RustRover

**Comment by Gnosnay** [comment]
The challenges I faced while reading about `bucket_sort` are as follows:  
- The control flow seems somewhat implicit, and I find it difficult to clearly understand the intention behind this part.  
- The documents are normalized within the `universe`, represented by a `RoaringBitmap`. This makes it challenging to easily infer the original document from the `docid` in the `universe`.  
- The concept of `Bucket` itself is a bit unclear to me, and I am not entirely sure what it refers to.  

I would greatly appreciate it if you could provide me with some guidance or clarification to help improve my understanding of the code.

**Comment by dureuill** [comment]
Hello,

thank you again for the time you put into investigating 👍 

I'll try to answer your questions:

- `bucket_sort` is about applying the ranking rules in the order they were defined to sort documents in buckets. A bucket in Meilisearch represents a group of documents that are considered the same in terms of relevancy (and hence share the same score).
- A ranking rule accepts as input a universe, that is, a list of all candidate documents for that rule, and returns a bucket, that is, the sublist of documents that is part of the universe and are the most relevant for that rule.
- From there, multiple things may happen:
    1. If this is the last ranking rule and the bucket is not empty, then remove the documents from the universe, and add them to results. We only actually add them to results if the documents are in the range that we want to return (that is controlled by `from` and `limit`. If we have enough documents, we also exit the algorithm altogether. If we don't, then we pop the last ranking rule and try again with the previous ranking rule.
    2. If this is not the last ranking rule and the bucket is not empty, then use the bucket as the universe for the next ranking rule. In a way, this creates "buckets in the bucket", and it further differentiates documents that were considered equal by the previous ranking rule.
    3. Whenever the bucket is empty for a ranking rule, go back to the previous ranking rule.
    4. If the bucket contains a single document, there's a sublety. If we don't care about the precise ranking score, then we can skip the remaining ranking rules, because they will always at most return the input document. However if we care about displaying the score (or using it e.g. in multi-search), then we should still compute the next ranking rules, because the score should not depend on whether or not there is a single document in the dataset matching the ranking rule. This is governed by the `scoring_strategy`.
- `RoaringBitmap`s are an efficient way of storing lists of `u32`. Internally, each document is assigned a u32 that we call an "internal document id" (in opposition to "external document id", which is set by the user). Meilisearch then uses lots of `RoaringBitmap` representing lists of documents, addressed by their internal docid. In particular the universe is the list of documents that can still be returned in buckets. If you manually indexed your documents and did not delete one of them, a heuristics that *currently* works to find the internal docid is to look at the order of insertion of the document: the first inserted document is 0, the second 1, etc.

Hope this clears up a bit what is going on here.
In summary, it appears that we might be going twice on the same ranking rule, due to a mistake in the code that selects the index of the current ranking rule in the vector of ranking rules?

**Comment by barloes** [comment]
Hi, i have added a pr detailing the changes.

https://github.com/meilisearch/meilisearch/pull/5313

current behavior when current_score < ranking_score_threshold -> back()!
- this would cause the cur_ranking_rule_index to point to already occupied index

suggest to add +1 to cur_ranking_rule_index after back()! so that cur_ranking_rule_index will point to not occupied index

what do you think?
tested with the new changes and return 1 item with 0.99 and 11 items with 0.98

**Comment by dureuill** [comment]
Closing this as it was fixed in v1.14 👍
